File: ios/Plugin/CapVolumePlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(CapVolumePlugin, "CapVolume",           
           CAP_PLUGIN_METHOD(setVolume, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(getVolume, CAPPluginReturnPromise);
)

File: ios/Plugin/CapVolumePlugin.swift
import Foundation
import Capacitor
import AVFoundation
import MediaPlayer

@objc(CapVolumePlugin)
public class CapVolumePlugin: CAPPlugin {
    private let implementation = CapVolume()

    @objc func setVolume(_ call: CAPPluginCall) {        
        let volume = call.getFloat("volume") ?? 1
        let volumeView = MPVolumeView(frame: .zero)
        volumeView.showsRouteButton = false
        let slider = volumeView.subviews.first(where: { $0 is UISlider }) as? UISlider
        slider?.setValue(volume, animated: false)
        call.resolve([
            "volume": implementation.echo(String(volume))
        ])
    }

    @objc func getVolume(_ call: CAPPluginCall) {
        let audioSession = AVAudioSession.sharedInstance()
        let currentVolume = audioSession.outputVolume
        call.resolve([
            "volume": implementation.echo(String(currentVolume))
        ])
    }
}
